server/src/main/java/com/exacaster/lighter/backend/yarn/YarnBackend.java

```diff
@@ -114,23 +111,26 @@ public void kill(Application application) {
                 "spark.yarn.appMasterEnv.PY_GATEWAY_HOST", host,
                 "spark.yarn.appMasterEnv.LIGHTER_SESSION_ID", application.getId()
         ));
-        if (yarnProperties.getKerberosKeytab() != null) {
+        if (yarnProperties.getKerberosKeytab() != null && yarnProperties.getKerberosPrincipal() != null) {
             props.put("spark.kerberos.keytab", yarnProperties.getKerberosKeytab());
             props.put("spark.kerberos.principal", yarnProperties.getKerberosPrincipal());
         }
         return props;
     }
 
     private Optional<String> getYarnApplicationId(Application application) {
+        var allStates = EnumSet.allOf(YarnApplicationState.class);
         return Optional.ofNullable(application.getAppId())
-                .or(() -> of(client.getApps(application.getId()))
-                        .map(YarnApplicationListResponse::getApps)
-                        .filter(Objects::nonNull)
-                        .map(YarnApplicationWrapper::getApp)
-                        .filter(Objects::nonNull)
-                        .flatMap(Collection::stream)
-                        .filter(Objects::nonNull)
-                        .max(Comparator.comparing(YarnApplication::getStartedTime))
-                        .map(YarnApplication::getId));
+                .or(() -> {
+                    try {
+                        return client.getApplications(Set.of("SPARK"), allStates, Set.of(application.getId())).stream()
```

Comment:
Maybe it would be nicer to use `getApplications` with `GetApplicationsRequest`, to avoid using allStates?